GH #1461: Enhancement - demixing modified to load movie stack in blocks

colleenjg opened the pull request:
# Overview:
`demixer.demix_time_dep_masks()` requires full movie stacks to be loaded,  
which can lead to out-of-memory (OOM) errors for large files.
(This pull request replaces the old, closed pull request #1464.)

# Addresses:
Addresses issue #1461

# Type of Fix:
New feature (non-breaking change which adds functionality)

# Solution:
`demixer.demix_time_dep_masks()` has been modified to accept a `block_size` 
parameter with a default value of 1000. Frames from the movie `stack` are now 
loaded and reshaped in blocks, avoiding OOM errors. 
`stack` can still be passed as an array as before, allowing backward compatibility, 
but now also as an hdf5 dataset. 

# Changes:
- `demixer.demix_time_dep_masks()` docstring updated
- `block_size` argument added with default value 1000
- `block_size` value of -1 loads stack in a single block
- Checks for invalid `block_size` values and raises an error if not -1 or >=1
- Movie stack reshaped within the demixing for loop, in blocks 
- Unit test `test_demixer.py` tests for `block_size` values -1 and 1000

# Validation:
Unit tests passed, as well as tests on local datasets.

# Notes:
This pull request replaces the old, closed pull request #1464.

isaak-willett commented on allensdk/test/brain_observatory/test_demixer.py:
```diff
@@ -62,17 +62,25 @@ def test_demix_raises_warning_for_singular_matrix(
 )
-def test_demix_time_dep_masks(raw_traces, stack, masks, expected):
-    result = dmx.demix_time_dep_masks(raw_traces, stack, masks)
+def test_demix_time_dep_masks(raw_traces, stack, masks, block_size, expected):
```
Add a test for block_size < 1 and not == -1, to test case where exception is raised.
You can break into another testing function and use
```python

 with pytest.raises(ValueError, match=<expected_message>):
    result = dmx.demix_time_dep_masks(raw_traces, stack, masks, block_size, expected)

```
I want to see this so that we can capture that exceptions are being raised when we expect them to be raised.

colleenjg replied:
Very good, will implement this.

isaak-willett commented on allensdk/brain_observatory/demixer.py:
```diff
@@ -120,19 +121,25 @@ def demix_time_dep_masks(raw_traces: np.ndarray, stack: np.ndarray,
-        stack = stack.reshape(T, P)
+    if block_size == -1:
+        block_size = T
+    elif block_size < 1:
```
also add exception checking for case when block_size is greater than T, ie: greater than the trace length
```python
elif block_size < 1 or block_size > T:

```

colleenjg replied:
The code currently handles this case without errors, however I realize it does not do so transparently.

I would propose not to throw an exception, for the following reason: if we do not want to constrain `block_size` to be a factor of `T` (I think this would be unnecessarily constraining), then the final block will almost always be shorter than `block_size`. The case where `block_size > T` is a special case of this, where the final block is the only block, and it is shorter than `block_size`.

Instead, I would propose that I implement these 3 changes to make this clearer.
1) **Rename `block_size` to `max_block_size`**: In the docstring, `block_size` is described as the **maximum** number of movie frames to read at a time. The modified variable name would be more transparent.
2) **Make stack indexing clearer**: [L156](https://github.com/colleenjg/AllenSDK/blob/GH-1461/feature/demixer-stack-block/allensdk/brain_observatory/demixer.py#L156) calculates the current block size, `block_T` (which is then used when reshaping the block on the next line). Currently, [L157](https://github.com/colleenjg/AllenSDK/blob/GH-1461/feature/demixer-stack-block/allensdk/brain_observatory/demixer.py#L157) loads the block using `stack[t : t+block_size]`. This indexing does work even when `t+block_size > T`, but it would be more transparent/cleaner to use `t+block_T` instead. 
**L157 change:**
\- `stack_block = stack[t : t+block_size].reshape(block_T, P)`
\+ `stack_block = stack[t : t+block_T].reshape(block_T, P)`
3) **Add test case**: Add a pytest case where `block_size > T`.

isaak-willett replied:
I agree with all these changes, thanks for putting thought into this! Once that's implemented I'll approve this.

colleenjg replied:
Thank you! I will push an updated commit ASAP.

colleenjg replied:
The changes have been made, and I squashed the commits.

isaak-willett commented:
@kschelonka since you've looked at this before I think it would be good for you to approve this as well before it get's merged.

colleenjg commented:
This latest commit reflects the modifications requested to date, squashed together.
In addition, I corrected an error I noticed in the docstring of `demixer.demix_time_dep_masks()`: the dimensions of the variable `raw_traces` were misstated as `(t, n)` instead of `(n, t)`.

colleenjg commented:
Hi @njmei,
Yes, the reshape is indeed the cause of the problem, as it forces the whole stack to be loaded into memory. I initially considered your solution, which does work. My concern was that it leads to the data being loaded into memory one row at a time. Considering that such files may be compressed/chunked, I thought it more optimal to read the data in chunks. 

Specifically, if `stack` is provided as an array, passing blocks to `_demix_point()` is just a view change, with no effect on memory usage. If `stack` is the movie file handle, then loading it into blocks can reduce the number of read/load operations by thousand-folds. In my tests on a smaller dataset (1 GB), this was marginally faster. However, I have not had a chance to do speed tests on, for example, a 60GB dataset. I do agree that my solution is bulkier, though.

Do let me know which solution you prefer, and I can push a new commit. I can also run a few speed tests on my end, as my assumptions may be incorrect.

njmei commented:
@colleenjg Great point! I wasn't considering HDF5, but should have! Reading in a block at a time should definitely be faster for larger datasets.
